### PR TITLE
add websitemonitor

### PR DESCRIPTION
**Pull request by mmartins-bonial**

### CODE REVIEW

File: site24x7/provider_test.go [hunk 1]
```diff
@@ -1,35 +1,27 @@
...
+//
+//func TestProvider_impl(t *testing.T) {
+//	var _ terraform.ResourceProvider = Provider()
+//}
+//
+//func testAccPreCheck(t *testing.T) {
+//	if v := os.Getenv("ZOHO_AUTHTOKEN"); v == "" {
+//		t.Fatal("ZOHO_AUTHTOKEN must be set for acceptance tests")
+//	}
+//}
```
Comment (martinohmann) [inline site24x7/provider_test.go]: Any reason you commented this?

File: site24x7/websitemonitor.go [hunk 2]
```diff
@@ -1,199 +1,230 @@
...
 )
 
+var WebsiteMonitorSchema = map[string]*schema.Schema{
+
+	"display_name": {
+		Type:     schema.TypeString,
+		Required: true,
+	},
+
+	"website": {
```
Comment (martinohmann) [inline site24x7/websitemonitor.go]: Can you remove the empty lines between the map items please? :)

File: site24x7/websitemonitor.go [hunk 3]
```diff
@@ -207,199 +238,72 @@ func websiteMonitorCreateOrUpdate(method, url string, expectedResponseStatus int
...
+		UnmatchingKeyword:     new(api.ValueAndSeverity),
+		MatchRegex:            new(api.ValueAndSeverity),
 		MatchCase:             d.Get("match_case").(bool),
 		UserAgent:             d.Get("user_agent").(string),
 		CustomHeaders:         customHeaders,
 		Timeout:               d.Get("timeout").(int),
 		LocationProfileID:     d.Get("location_profile_id").(string),
-		NotificationProfileID: d.Get("notification_profile_id").(string),
-		ThresholdProfileID:    d.Get("threshold_profile_id").(string),
+		NotificationProfileID: d.Get("NotificationProfileID").(string),
```
Comment (martinohmann) [inline site24x7/websitemonitor.go]: I think you want `"notification_profile_id"` here :).

File: site24x7/websitemonitor.go [hunk 4]
```diff
@@ -207,199 +238,72 @@ func websiteMonitorCreateOrUpdate(method, url string, expectedResponseStatus int
...
+		ThresholdProfileID:    d.Get("ThresholdProfileID").(string),
```
Comment (martinohmann) [inline site24x7/websitemonitor.go]: I think you want `"threshold_profile_id"` here.

File: site24x7/websitemonitor.go [hunk 5]
```diff
@@ -207,199 +238,72 @@ func websiteMonitorCreateOrUpdate(method, url string, expectedResponseStatus int
...
-		d.Set("threshold_profile_id", id)
-	}
-	if len(m.UserGroupIDs) == 0 {
-		id, err := defaultUserGroup(client)
-		if err != nil {
-			return err
-		}
-		m.UserGroupIDs = []string{id}
-		d.Set("user_group_ids", []string{id})
-	}
```
Comment (martinohmann) [inline site24x7/websitemonitor.go]: The code you removed between line 244 and 295 is actually necessary and can be used almost as is. Basically you need to replace the calls to the `default*` funcs with the new `Default*` funcs we created in the previous PR. The original code can also be simplified a bit.

File: site24x7/websitemonitor.go [hunk 6]
```diff
@@ -207,199 +238,72 @@ func websiteMonitorCreateOrUpdate(method, url string, expectedResponseStatus int
...
 		CheckFrequency:        strconv.Itoa(d.Get("check_frequency").(int)),
 		HTTPMethod:            d.Get("http_method").(string),
 		AuthUser:              d.Get("auth_user").(string),
 		AuthPass:              d.Get("auth_pass").(string),
-		MatchingKeyword:       new(ValueAndSeverity),
-		UnmatchingKeyword:     new(ValueAndSeverity),
-		MatchRegex:            new(ValueAndSeverity),
+		MatchingKeyword:       new(api.ValueAndSeverity),
+		UnmatchingKeyword:     new(api.ValueAndSeverity),
+		MatchRegex:            new(api.ValueAndSeverity),
```
Comment (martinohmann) [inline site24x7/websitemonitor.go]: You can remove lines 261-263 as the initialize the fields with `nil` which is the zero value anyways.

File: site24x7/websitemonitor.go [hunk 7]
```diff
@@ -207,199 +238,72 @@ func websiteMonitorCreateOrUpdate(method, url string, expectedResponseStatus int
 	}
 
 	i := 0
-	actionRefs := make([]ActionRef, len(d.Get("actions").(map[string]interface{})))
-	for k, v := range d.Get("actions").(map[string]interface{}) {
+	actionRefs := make([]api.ActionRef, len(d.Get("action_ids").(map[string]interface{})))
+	for k, v := range d.Get("action_ids").(map[string]interface{}) {
 		tmp, err := strconv.Atoi(k)
 		if err != nil {
 			panic(err)
```
Comment (martinohmann) [inline site24x7/websitemonitor.go]: We should not panic here. This makes the code fragile and hard to test. Make the `resourceDataToWebsiteMonitor` func return an error as the 2nd return value instead and pass it through up the stack.

The `panic` is also the reason why your tests were crashing.

File: site24x7/provider_test.go (hunk 1 above)
Comment (mmartins-bonial) [reply site24x7/provider_test.go]: This is not needed, I will comment it in again

File: site24x7/websitemonitor.go (hunk 2 above)
Comment (mmartins-bonial) [reply site24x7/websitemonitor.go]: Done

File: site24x7/websitemonitor.go (hunk 3 above)
Comment (mmartins-bonial) [reply site24x7/websitemonitor.go]: Done

File: site24x7/websitemonitor.go (hunk 4 above)
Comment (mmartins-bonial) [reply site24x7/websitemonitor.go]: Done

File: site24x7/websitemonitor.go (hunk 6 above)
Comment (mmartins-bonial) [reply site24x7/websitemonitor.go]: Done

File: site24x7/provider_test.go (hunk 1 above)
Comment (martinohmann) [reply site24x7/provider_test.go]: Then just leave it there. We will address that in another PR as is does not have anything to do with the feature you're implementing in this PR ;).

**Pull request merged**